# 1. Lection 1 - 08. Mar. 2018
## 1.1. Evolutionary Computation

What is evolutionary computing:
https://en.wikipedia.org/wiki/Evolutionary_computation

Confidence -> p. values, C-Intervals

Computational intelligence emerged out of big data.

Heuristic algorithm: cant always give solution, if solution you dont know if it is useable or correct

Computational Intelligence vs AI:

Both fall under maschine learning

AI -> hard computing -> always the same algorithm -> like a monkey 
CI -> soft computing (soft boundaries, fuzzy logic(next semester)) -> very adaptive -> more complex -> makes it heuristic

Clinical Trial, Beta Roll out, tweaking, market.

Patient in the ICU -> artificial breathing -> wening (research that) -> liung machine gradually needs to lower its power so the lungs start breathing themselves again.

MYCT->Processsor Speed

## 1. Styles of learning

Classification:
100 records, 80 records used to trained the classifier ()training set, this classifier is tested against the remaining 20 (test set).


Association:
combining classifiers

Clustering:



## Attributes



## I/O Problems

Choose Appropriate architectures -> disadvantages: you dont know if the other algo would have worked

Convert input -> disadvantages: accuracy of data gets lost and conversion rules need to be made


## Conversion

Hamming Distance

# First Execrise:

I decided to use Jupyter Lab which is a self hosted development web ui for scientific note taking and data visualization. It is based on the open source project jupyter notebook also known as ipython shell.

Besides python it supports multiple kernels which can be installed as you need them.

For this learning journal I installed iRkernel to solve the R execises.

Jupyter Lab feels like a traditional IDE with multiple configurable panels and splitable views.

You have a file tree on the left, the main code panel for the notebook on the right. The Notebook itself consists of cells which can also have different kernels active. This enables to use Markdown in one cell and R in the next one. Markdown and raw text interpretation are shipped by default.

https://jupyterlab.readthedocs.io/en/stable/getting_started/overview.html


After installing Jupyter Lab and trying the first execise with rweka i had to run ```sudo R CMD javareconf``` and add some jvm shared libraries to the `LD_LIBRARY_PATH`

This might also only have been necessary because the linux distribution I am running.

--------------


In [2]:
#install.packages("RWeka")

library(RWeka)

#Assignment 1
mushrooms <- read.csv("mushrooms.csv", stringsAsFactors = TRUE)

str(mushrooms)

'data.frame':	8124 obs. of  23 variables:
 $ type                    : Factor w/ 2 levels "edible","poisonous": 2 1 1 2 1 1 1 1 2 1 ...
 $ cap_shape               : Factor w/ 6 levels "bell","conical",..: 3 3 1 3 3 3 1 1 3 1 ...
 $ cap_surface             : Factor w/ 4 levels "fibrous","grooves",..: 4 4 4 3 4 3 4 3 3 4 ...
 $ cap_color               : Factor w/ 10 levels "brown","buff",..: 1 10 9 9 4 10 9 9 9 10 ...
 $ bruises                 : Factor w/ 2 levels "no","yes": 2 2 2 2 1 2 2 2 2 2 ...
 $ odor                    : Factor w/ 9 levels "almond","anise",..: 8 1 2 8 7 1 1 2 8 1 ...
 $ gill_attachment         : Factor w/ 2 levels "attached","free": 2 2 2 2 2 2 2 2 2 2 ...
 $ gill_spacing            : Factor w/ 2 levels "close","crowded": 1 1 1 1 2 1 1 1 1 1 ...
 $ gill_size               : Factor w/ 2 levels "broad","narrow": 2 1 1 2 1 1 1 1 2 1 ...
 $ gill_color              : Factor w/ 12 levels "black","brown",..: 1 1 2 2 1 2 5 2 8 5 ...
 $ stalk_shape             : Factor w/

## Assignment 1:

### Take a look at the database, and describe it in your learning journal

Dataset has 23 variables with 2-12 levels. 

The variables describe the appearance and behaviour of the mushrooms. 

"Type" for example has to levels -> edible and poisonous while cap color is much more granulous with 10 distinguished colors for the cap of the mushroom.

The variables describe the appearance, behaviour, odor, spatial distribution and so on.

-------------------------------------

## Assignment 2

### use the str() command on all 23 variables. Do you notice anything unusual?


In [14]:
#Assignment 2
for ( i in 1:length(mushrooms) ) {
    str(mushrooms[[i]])
}

 Factor w/ 2 levels "edible","poisonous": 2 1 1 2 1 1 1 1 2 1 ...
 Factor w/ 6 levels "bell","conical",..: 3 3 1 3 3 3 1 1 3 1 ...
 Factor w/ 4 levels "fibrous","grooves",..: 4 4 4 3 4 3 4 3 3 4 ...
 Factor w/ 10 levels "brown","buff",..: 1 10 9 9 4 10 9 9 9 10 ...
 Factor w/ 2 levels "no","yes": 2 2 2 2 1 2 2 2 2 2 ...
 Factor w/ 9 levels "almond","anise",..: 8 1 2 8 7 1 1 2 8 1 ...
 Factor w/ 2 levels "attached","free": 2 2 2 2 2 2 2 2 2 2 ...
 Factor w/ 2 levels "close","crowded": 1 1 1 1 2 1 1 1 1 1 ...
 Factor w/ 2 levels "broad","narrow": 2 1 1 2 1 1 1 1 2 1 ...
 Factor w/ 12 levels "black","brown",..: 1 1 2 2 1 2 5 2 8 5 ...
 Factor w/ 2 levels "enlarging","tapering": 1 1 1 1 2 1 1 1 1 1 ...
 Factor w/ 5 levels "bulbous","club",..: 3 2 2 3 3 2 2 2 3 2 ...
 Factor w/ 4 levels "fibrous","scaly",..: 4 4 4 4 4 4 4 4 4 4 ...
 Factor w/ 4 levels "fibrous","scaly",..: 4 4 4 4 4 4 4 4 4 4 ...
 Factor w/ 9 levels "brown","buff",..: 8 8 8 8 8 8 8 8 8 8 ...
 Factor w/ 9 levels "brown","buf

After using the `str()` function on all the variables it shows that using `str()` on the dataframe it describes the dataframe itself displaying the object type, the total amount of objects, the variables. 

When using `str()` on all the variables separately it doesn't show the variable names anymore and only the content type and head of content. 

-------------------------------

## Assignment 3
###  Record the output and your analysis of the output in your learning journal. How does the classifier perform? 

In [27]:
#Assignment 3
mushroom_1r <- OneR(type~., data=mushrooms)
mushroom_1r

odor:
	almond	-> edible
	anise	-> edible
	creosote	-> poisonous
	fishy	-> poisonous
	foul	-> poisonous
	musty	-> poisonous
	none	-> edible
	pungent	-> poisonous
	spicy	-> poisonous
(8004/8124 instances correct)


The classifier got 8004 out of 8124 instances correct. I think it performed quite well.

-----------------------

## Assignment 4
### Analyze the Summary and Confusion Matrix printouts. Are there any terms that you do not understand? Look these up in google, and record their meaning. 

### Also, give an interpretation of the confusion matrix. How did the classifier perform?

In [28]:
#Assignment 4
summary(mushroom_1r)


=== Summary ===

Correctly Classified Instances        8004               98.5229 %
Incorrectly Classified Instances       120                1.4771 %
Kappa statistic                          0.9704
Mean absolute error                      0.0148
Root mean squared error                  0.1215
Relative absolute error                  2.958  %
Root relative squared error             24.323  %
Total Number of Instances             8124     

=== Confusion Matrix ===

    a    b   <-- classified as
 4208    0 |    a = edible
  120 3796 |    b = poisonous

120 instances were classfied incorrectly as edible which is bad for the mushroom gatherer, he poisoned now :'(

All poisonous instances have been correctly classified as poisonous.


Unknown terms:

- Kappa statistics: value to mesaure interrater agreement which describes the degree of agreement between to parties.

-----------------------------

## ASSIGNMENT 5: Analyze the Summary and Confusion Matrix printouts again. Did this classifier perform better? If yes, how so?


In [26]:
mushroom_JRip <- JRip(type ~ ., data = mushrooms)
mushroom_JRip
summary(mushroom_JRip)

JRIP rules:

(odor = foul) => type=poisonous (2160.0/0.0)
(gill_size = narrow) and (gill_color = buff) => type=poisonous (1152.0/0.0)
(gill_size = narrow) and (odor = pungent) => type=poisonous (256.0/0.0)
(odor = creosote) => type=poisonous (192.0/0.0)
(spore_print_color = green) => type=poisonous (72.0/0.0)
(stalk_surface_below_ring = scaly) and (stalk_surface_above_ring = silky) => type=poisonous (68.0/0.0)
(habitat = leaves) and (gill_attachment = free) and (population = clustered) => type=poisonous (16.0/0.0)
 => type=edible (4208.0/0.0)

Number of Rules : 8



=== Summary ===

Correctly Classified Instances        8124              100      %
Incorrectly Classified Instances         0                0      %
Kappa statistic                          1     
Mean absolute error                      0     
Root mean squared error                  0     
Relative absolute error                  0      %
Root relative squared error              0      %
Total Number of Instances             8124     

=== Confusion Matrix ===

    a    b   <-- classified as
 4208    0 |    a = edible
    0 3916 |    b = poisonous

Using JRip all instances have been classified correctly.

In the previously used OneR classifier poisonousness was determined by the different odors only.

JRip determined more rules that not only described one condition, but also combinations of features in boolean expressions.


# Second Exercise

## Neural networks in Healthcare


### Predicting cancer outcomes from histology and genomics using convolutional networks
#### http://www.pnas.org/content/115/13/E2970


# TODO

-----------------------

# Third Exercise

## Neural networks in R


loaded the concrete.csv into variable using read.csv from library RWeka.

look at it using str()


In [60]:
concrete <- read.csv("concrete.csv")

print("describe structure of concrete and its variables using str()")
str(concrete)
for ( i in 1:length(concrete) ) {
    str(concrete[[i]])
}



[1] "describe structure of concrete and its variables using str()"
'data.frame':	1030 obs. of  9 variables:
 $ cement      : num  141 169 250 266 155 ...
 $ slag        : num  212 42.2 0 114 183.4 ...
 $ ash         : num  0 124.3 95.7 0 0 ...
 $ water       : num  204 158 187 228 193 ...
 $ superplastic: num  0 10.8 5.5 0 9.1 0 0 6.4 0 9 ...
 $ coarseagg   : num  972 1081 957 932 1047 ...
 $ fineagg     : num  748 796 861 670 697 ...
 $ age         : int  28 14 28 28 28 90 7 56 28 28 ...
 $ strength    : num  29.9 23.5 29.2 45.9 18.3 ...
 num [1:1030] 141 169 250 266 155 ...
 num [1:1030] 212 42.2 0 114 183.4 ...
 num [1:1030] 0 124.3 95.7 0 0 ...
 num [1:1030] 204 158 187 228 193 ...
 num [1:1030] 0 10.8 5.5 0 9.1 0 0 6.4 0 9 ...
 num [1:1030] 972 1081 957 932 1047 ...
 num [1:1030] 748 796 861 670 697 ...
 int [1:1030] 28 14 28 28 28 90 7 56 28 28 ...
 num [1:1030] 29.9 23.5 29.2 45.9 18.3 ...


### ASSIGNMENT 1: Take a look at the database, and describe it in your learning journal. 

9 Variables, 1030 Objects, only numerical values

In [70]:
print("Cement:")
head(table(concrete$cement))
print("Slag:")
head(table(concrete$slag))
print("Ash:")
head(table(concrete$ash))
print("Water:")
head(table(concrete$water))
print("Superplastic:")
head(table(concrete$superplastic))
print("Coarseagg:")
head(table(concrete$coarseagg))
print("fineagg:")
head(table(concrete$fineagg))
print("age:")
head(table(concrete$age))
print("strength:")
head(table(concrete$strength))

[1] "Cement:"



  102 108.3   116 122.6   132   133 
    4     4     4     4     2     5 

[1] "Slag:"



   0   11 13.6   15 17.2 17.5 
 471    4    5    5    1    1 

[1] "Ash:"



   0 24.5   59   60   71 71.5 
 566   15    1    1    1    1 

[1] "Water:"



121.8 126.6   127 127.3 137.8   140 
    5     5     1     1     5     1 

[1] "Superplastic:"



  0 1.7 1.9   2 2.2 2.5 
379   4   1   1   1   2 

[1] "Coarseagg:"



  801 801.1 801.4   811   814 814.1 
    4     1     1     2     1     1 

[1] "fineagg:"



  594   605 611.8   612   613 613.2 
   30     5     5     1    22     2 

[1] "age:"



  1   3   7  14  28  56 
  2 134 126  62 425  91 

[1] "strength:"



2.33 3.32 4.57 4.78 4.83  4.9 
   1    1    1    1    1    1 

### Assignemnt 2

Problem here is that the values are not between one and 0, therefore not being normalized.
In neural networks values are activated using activation functions like the very basic sigmoid function or the ReLU (Rectified Linear Unit) function.

In [76]:
normalize <- function(x) {
 return((x - min(x)) / (max(x) - min(x)))
}

concrete_norm <- as.data.frame(lapply(concrete, normalize))

summary(concrete_norm$strength)
summary(concrete$strength)

concrete_train <- concrete_norm[1:773, ]
concrete_test <- concrete_norm[774:1030, ]

print("summary of concrete_train:")
summary(concrete_train)

print("summary of concrete_test")
summary(concrete_test)

   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
 0.0000  0.2664  0.4001  0.4172  0.5457  1.0000 

   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   2.33   23.71   34.45   35.82   46.13   82.60 

[1] "summary of concrete_train:"


     cement            slag              ash             water       
 Min.   :0.0000   Min.   :0.00000   Min.   :0.0000   Min.   :0.0000  
 1st Qu.:0.2055   1st Qu.:0.00000   1st Qu.:0.0000   1st Qu.:0.3450  
 Median :0.3628   Median :0.05565   Median :0.0000   Median :0.5104  
 Mean   :0.4055   Mean   :0.20589   Mean   :0.2627   Mean   :0.4824  
 3rd Qu.:0.5662   3rd Qu.:0.39733   3rd Qu.:0.5912   3rd Qu.:0.5679  
 Max.   :1.0000   Max.   :0.95186   Max.   :0.9995   Max.   :1.0000  
  superplastic      coarseagg         fineagg            age         
 Min.   :0.0000   Min.   :0.0000   Min.   :0.0000   Min.   :0.00000  
 1st Qu.:0.0000   1st Qu.:0.3808   1st Qu.:0.3352   1st Qu.:0.03571  
 Median :0.1863   Median :0.4855   Median :0.4629   Median :0.07418  
 Mean   :0.1851   Mean   :0.5067   Mean   :0.4504   Mean   :0.12523  
 3rd Qu.:0.3137   3rd Qu.:0.6657   3rd Qu.:0.5795   3rd Qu.:0.15110  
 Max.   :1.0000   Max.   :1.0000   Max.   :1.0000   Max.   :1.00000  
    strength     
 M

[1] "summary of concrete_test"


     cement             slag              ash             water       
 Min.   :0.01438   Min.   :0.00000   Min.   :0.0000   Min.   :0.0000  
 1st Qu.:0.22374   1st Qu.:0.00000   1st Qu.:0.0000   1st Qu.:0.3291  
 Median :0.42466   Median :0.06678   Median :0.0000   Median :0.4617  
 Mean   :0.41973   Mean   :0.20475   Mean   :0.2952   Mean   :0.4623  
 3rd Qu.:0.58676   3rd Qu.:0.40122   3rd Qu.:0.5927   3rd Qu.:0.5607  
 Max.   :1.00000   Max.   :1.00000   Max.   :1.0000   Max.   :0.9177  
  superplastic      coarseagg         fineagg            age         
 Min.   :0.0000   Min.   :0.0000   Min.   :0.0000   Min.   :0.00000  
 1st Qu.:0.0000   1st Qu.:0.3674   1st Qu.:0.3588   1st Qu.:0.01648  
 Median :0.2422   Median :0.4826   Median :0.4681   Median :0.07418  
 Mean   :0.2156   Mean   :0.4790   Mean   :0.4510   Mean   :0.11509  
 3rd Qu.:0.3261   3rd Qu.:0.6448   3rd Qu.:0.5725   3rd Qu.:0.07418  
 Max.   :1.0000   Max.   :0.9689   Max.   :1.0000   Max.   :1.00000  
    strength 

### Assignment 3 Compare summary of training and test data sets
In both data sets the minimal and maximal values are exactly or at least close to 0 and 1. whiche lets assume the data is quite evenly distributed.

Also the other statistical key figures like 1st quartile, median, mean and 3rd quartile are similar.

--------------------------------

In [102]:
library(neuralnet)

concrete_model <- neuralnet(strength ~ cement + slag + ash + water +
superplastic + coarseagg + fineagg + age, data = concrete_train)
plot(concrete_model)
#TODO plot not showing

### Assignment 4 describe the plot

### Biological Neural Networks 10.04.2018

research: Parametrisch

Can we simulate the Human Brain Functions? SOFA Framework?

Euclidean distance: w.t.h. ?

### 11.04.2018

Fixe Daten können als Bias Nodes verwendet werden (research that!)



### two publications from pubmed

What, medical background
what method, what neural network
compare both publications

pubmed search string: artificial neural networks.

found articles: 
#### Predicting cancer outcomes from histology and genomics using convolutional networks.
https://www.ncbi.nlm.nih.gov/pubmed/29531073

#### Deep Learning for Drug Design: an Artificial Intelligence Paradigm for Drug Discovery in the Big Data Era.
https://www.ncbi.nlm.nih.gov/pubmed/29603063



### Backward Propagation 24.4.2018

output is the squashing function

How do I know what a squashing function is?
for example do I know that f(x)=1/(1+e^x)

dEtotal/dw7=dEtotal/douto2*douto2/dneto2*dneto2/dw7

### OCR 03.05.2018

# TODO ASSIGNMENT 1: Take a look at the database, and describe it.




In [ ]:
letters <- read.csv("letterdata.csv")
head(letters)

letters_train <- letters[1:16000, ]
letters_test <- letters[16001:20000, ]

#install.packages("kernlab") #only needs to be installed once
library(kernlab)

letter_classifier <- ksvm(letter ~ ., data = letters_train, kernel = "vanilladot")

letter_classifier



# TODO Assignment 2




In [ ]:
letter_predictions <- predict(letter_classifier, letters_test)
head(letter_predictions)

table(letter_predictions, letters_test$letter)
agreement <- letter_predictions == letters_test$letter
table(agreement)

In [ ]:
letter_classifier_rbf <- ksvm(letter ~ ., data = letters_train, kernel = "rbfdot")
letter_predictions_rbf <- predict(letter_classifier_rbf, letters_test)
head(letter_predictions_rbf)

table(letter_predictions_rbf, letters_test$letter)
agreement_rbf <- letter_predictions_rbf == letters_test$letter
table(agreement_rbf)

In [ ]:
letter_classifier_polydot <- ksvm(letter ~ ., data = letters_train, kernel = "polydot")
letter_predictions_polydot <- predict(letter_classifier_polydot, letters_test)
head(letter_predictions_polydot)

table(letter_predictions_polydot, letters_test$letter)
agreement_polydot <- letter_predictions_polydot == letters_test$letter
table(agreement_polydot)

In [ ]:
letter_classifier_tanhdot <- ksvm(letter ~ ., data = letters_train, kernel = "tanhdot")
letter_predictions_tanhdot <- predict(letter_classifier_tanhdot, letters_test)
head(letter_predictions_tanhdot)

table(letter_predictions_tanhdot, letters_test$letter)
agreement_tanhdot <- letter_predictions_tanhdot == letters_test$letter
table(agreement_tanhdot)

### Hierarchical mixture of experts

4-6 Folien 5-10 min Präsentation

Hierarchical mixture of experts: non-linear combination of individual outputs by a multiple gating nodes arranged in a hierarchical fashion. bis Mittwoch 23. Mai.

### Exercise Bagging 22.05.2018

Assignment 1


In [ ]:
credit <- read.csv("credit.csv")
#str(credit)
table(credit$checking_balance) 
table(credit$savings_balance)
summary(credit$months_loan_duration)
summary(credit$amount)

#### Assignment 2 
firstly I had to again install the ipred package


In [ ]:
library(ipred)

set.seed(300)
mybag <- bagging(default ~ ., data = credit, nbagg = 25)
credit_pred <- predict(mybag, credit)
table(credit_pred, credit$default)

library(caret) # install it first using install.packages("caret") 
#I also needed to install the "gcc-fortran" package for my arch linux installation, might be "gfortran" in debian based distros

set.seed(300)
ctrl <- trainControl(method = "cv", number = 10)
train(default ~ ., data = credit, method = "treebag", trControl = ctrl) #first time package "e1071" was missing so i installed that in R.

#### Explain last code

In [ ]:
str(svmBag)
bagctrl <- bagControl(fit = svmBag$fit, predict = svmBag$pred, aggregate = svmBag$aggregate)
set.seed(300)
svmbag <- train(default ~ ., data = credit, "bag", trControl = ctrl, bagControl = bagctrl)
svmbag
# for further research: https://cran.r-project.org/web/packages/caret/caret.pdf

### 05.06.2018 TicTacToe

Image preprocessing:
imager autocrop: https://www.rdocumentation.org/packages/imager/versions/0.41.1/topics/autocrop

magick transform with threshold for straightening: 

https://www.rdocumentation.org/packages/magick/versions/1.9/topics/transform

library(magick)
library(tesseract)
im <- image_read("ttt8.jpeg")

#library(magrittr)
#library(imager)

#autocrop(im, color = color.at(im, 20, 20), axes = "yx")

im <- image_trim(im, fuzz=20)
im <- image_resize(im, "300x300!")
im <- image_contrast(im)
im <- image_contrast(im)
im <- image_transparent(im, "black", fuzz=30)
im <- image_crop(im, "100x100+10")
im

height <- 0
width <- 0

l <- c()
#cat(l)

while ( height<=200) {
    #cat("H: ",height,"\n")
    while (width<=200) {
        #cat("W: ",width,"\n")
        cropstr <- capture.output(cat(height,"x",width))
        print(cropstr)
        im <- image_crop(im, cropstr)
        #print(im)
        l<-c(l,im)
        width=width+100
        
    }
    height=height+100
    width=0
}

cat(image_ocr(im))

result_v <- vector(,9)


prepare_for_O <- function(ocr_sample_input) {
 replacements_O <- c("o", "0", "g", "G", "ö", "Ö", "C", "c", "q", "Q")
 
    didReplace <- FALSE
    
 updated_set_O <- c() 
 for (input in ocr_sample_input) {
  for (r_O in replacements_O) {
   if (input == r_O) {
    updated_set_O <- c(updated_set_O, "O")
       didReplace  <- TRUE
       break
   }
  }
     if (!didReplace) {
         updated_set_O <- c(updated_set_O, input)
     }
     didReplace <- FALSE
 }
 return(updated_set_O)
}

ocr_sample_input = c("X", "O", "X", "O", "X", "c", "G", "C", "M")

prepare_for_X <- function(ocr_sample_input) {
 replacements_X <- c("x", "X", "m", "M" ,"k", "K", "z", "Z")
 
    didReplace <- FALSE
    
 updated_set_X <- c() 
 for (input in ocr_sample_input) {
  for (r_X in replacements_X) {
   if (input == r_X) {
    updated_set_X <- c(updated_set_X, "X")
       didReplace  <- TRUE
       break
   }
  }
     if (!didReplace) {
         updated_set_X <- c(updated_set_X, input)
     }
     didReplace <- FALSE
 }
 return(updated_set_X)
}


positions <- prepare_for_X(prepare_for_O(ocr_sample_input))

comparedata <- c(c(1,2,3),c(4,5,6),c(7,8,9),c(1,4,7),c(2,5,8),c(3,6,9),c(1,5,4),c(3,5,7))


indicesFor <- function(val, inputArray) {
 indices <- c()

 index <- 1

for (input in inputArray) {
  if (input == val) {
    indices <- c(indices, index)
  }
   index <- index + 1
  }
return(indices)
}

x_positions <- indicesFor("X", positions)

comparedata[6][1] %in% x_positions
#right[1] %in% x_positions

x_positions

for (right in comparedata){
    result1 <- right[1] %in% x_positions
    result2 <- right[2] %in% x_positions
    result3 <- right[3] %in% x_positions
    #print(right[1] %in% x_positions)
    #print(result1 & result2)
    if(result1 & result2 & result3){
        print("\n\n\nyay\n\n\n")
        #print(right)
    }
}


#for (tile in )

#cat(image_ocr(im2))
#image_deskew(im2, threshold = 40)
#tesseract::ocr

Pürfungsrelevant
Squashing functions
What

SVM (keine vectorrechnung)

(Kein mathe bei radial bas** functions)

Perceptron, Backwardpropagation

OCR

Bagging Boosting (eigene Präsentationen und die der Kollegen)

Nichts zum Tic Tac Toe


## Exporting the learning journal

I had to adapt the underlying latex template thats used for exporting the notebook to pdf.

The resource I used for that was this url: http://www.markus-beuckelmann.de/blog/customizing-nbconvert-pdf.html

Basically the code input and output cells were not wrapped at all so a wrapping function needed to be added.

This worked better:
http://compbio.ucsd.edu/outputting-beautiful-jupyter-notebooks-r-kernel-edition/